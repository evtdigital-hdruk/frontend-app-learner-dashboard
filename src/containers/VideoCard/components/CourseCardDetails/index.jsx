import React from 'react';
import PropTypes from 'prop-types';

import { Button } from '@openedx/paragon';

import useCardDetailsData from './hooks';
import './index.scss';

export const CourseCardDetails = ({ cardId }) => {
  const {
    accessMessage,
    isEntitlement,
    isFulfilled,
    canChange,
    openSessionModal,
    changeOrLeaveSessionMessage,
  } = useCardDetailsData({ cardId });

  return (
    <span className="small" data-testid="CourseCardDetails">
      {!(isEntitlement && !isFulfilled) && accessMessage && (
        ` • ${accessMessage}`
      )}
      {isEntitlement && isFulfilled && canChange ? (
        <>
          {' • '}
          <Button variant="link" size="inline" className="m-0 p-0" onClick={openSessionModal}>
            {changeOrLeaveSessionMessage}
          </Button>
        </>
      ) : null}
    </span>
  );
};

CourseCardDetails.propTypes = {
  cardId: PropTypes.string.isRequired,
};

CourseCardDetails.defaultProps = {};

export default CourseCardDetails;
