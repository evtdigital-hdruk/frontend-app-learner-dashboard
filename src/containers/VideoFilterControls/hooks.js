import React from 'react';
import { useToggle } from '@openedx/paragon';

import { StrictDict } from 'utils';
import track from 'tracking';
import { reduxHooks } from 'hooks';

import * as module from './hooks';

export const state = StrictDict({
  target: (val) => React.useState(val), // eslint-disable-line
});

/**
 * Sets up a toggle for the modal as well as helper functions for handling changes to the form controls.
 *
 * @param {array} filters Currently active video filters
 * @param {function} setSortBy Set function for sorting the video list
 * @returns {object} data and functions for managing the VideoFilterControls component
 */
export const useVideoFilterControlsData = ({
  filters,
  setSortBy,
}) => {
  const [isOpen, toggleOpen, toggleClose] = useToggle(false);
  const [target, setTarget] = module.state.target(null);

  const addFilter = reduxHooks.useAddVideoFilter();
  const removeFilter = reduxHooks.useRemoveVideoFilter();

  const handleFilterChange = ({ target: { checked, value } }) => {
    const update = checked ? addFilter : removeFilter;
    update(value);
  };
  const handleSortChange = ({ target: { value } }) => {
    setSortBy(value);
  };

  const open = () => {
    track.filter.filterClicked();
    toggleOpen();
  };

  const close = () => {
    track.filter.filterOptionSelected(filters);
    toggleClose();
  };

  return {
    isOpen,
    open,
    close,
    target,
    setTarget,
    handleFilterChange,
    handleSortChange,
  };
};

export default useVideoFilterControlsData;
