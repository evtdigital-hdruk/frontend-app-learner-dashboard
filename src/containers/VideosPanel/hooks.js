import React from 'react';

import queryString from 'query-string';

import { ListPageSize, SortKeys } from 'data/constants/app';
import { reduxHooks } from 'hooks';
import { StrictDict } from 'utils';

import * as module from './hooks';

export const state = StrictDict({
  sortBy: (val) => React.useState(val), // eslint-disable-line
});

/**
 * Filters are fetched from the store and used to generate a list of "visible" courses.
 * Other values returned and used for the layout of the CoursesPanel component are:
 * the current page number, the sorting method, and whether or not to enable filters and pagination.
 *
 * @returns data for the CoursesPanel component
 */
export const useVideoListData = () => {
  const videoFilters = reduxHooks.useVideoFilters();
  const removeVideoFilter = reduxHooks.useRemoveVideoFilter();
  const videoPageNumber = reduxHooks.useVideoPageNumber();
  const setVideoPageNumber = reduxHooks.useSetVideoPageNumber();

  const [sortBy, setSortBy] = module.state.sortBy(SortKeys.enrolled);

  const querySearch = queryString.parse(window.location.search, { parseNumbers: true });

  const { numPages, visibleVideoList } = reduxHooks.useCurrentVideoList({
    sortBy,
    videoFilters,
    pageSize: querySearch?.disable_pagination === 1 ? 0 : ListPageSize,
  });

  const handleRemoveFilter = (filter) => () => removeVideoFilter(filter);

  return {
    videoPageNumber,
    numPages,
    setVideoPageNumber,
    visibleVideoList,
    filterOptions: {
      sortBy,
      setSortBy,
      videoFilters,
      handleRemoveFilter,
    },
    showFilters: videoFilters.length > 0,
  };
};

export default useVideoListData;
