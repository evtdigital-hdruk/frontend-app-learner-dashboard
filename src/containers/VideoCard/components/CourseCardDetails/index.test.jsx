import React from 'react';
import { shallow } from 'enzyme';

import CourseCardDetails from '.';

import hooks from './hooks';

jest.mock('./hooks', () => ({
  __esModule: true,
  default: jest.fn(),
}));

const cardId = 'test-card-id';

describe('CourseCard Details component', () => {
  const defaultHooks = {
    providerName: 'provider-name',
    accessMessage: 'access-message',
    openSessionModal: jest.fn().mockName('useSelectSession.openSessionModal'),
    isEntitlement: true,
    isFulfilled: true,
    canChange: true,
    courseNumber: 'test-course-number',
    changeOrLeaveSessionMessage: 'change-or-leave-session-message',
  };
  const createWrapper = (hookOverrides = {}) => {
    hooks.mockReturnValueOnce({
      ...defaultHooks,
      ...hookOverrides,
    });
    return shallow(<CourseCardDetails cardId={cardId} />);
  };

  test('has change session button on entitlement course', () => {
    const wrapper = createWrapper();
    expect(wrapper).toMatchSnapshot();
    // it has 3 separator, 4 column
    expect(wrapper.text().match(/•/g)).toHaveLength(3);
  });

  test('has change session button on entitlement course but no access message', () => {
    const wrapper = createWrapper({ accessMessage: null });
    expect(wrapper).toMatchSnapshot();
    // it has 2 separator, 3 column
    expect(wrapper.text().match(/•/g)).toHaveLength(2);
  });

  test('does not have change session button on regular course', () => {
    const wrapper = createWrapper({ isEntitlement: false });
    expect(wrapper).toMatchSnapshot();
    // it has 2 separator, 3 column
    expect(wrapper.text().match(/•/g)).toHaveLength(2);
  });
});
