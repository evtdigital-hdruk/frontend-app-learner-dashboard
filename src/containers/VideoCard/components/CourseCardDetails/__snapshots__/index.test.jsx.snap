// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseCard Details component does not have change session button on regular course 1`] = `
<span
  className="small"
  data-testid="CourseCardDetails"
>
  provider-name
   • 
  test-course-number
   • access-message
</span>
`;

exports[`CourseCard Details component has change session button on entitlement course 1`] = `
<span
  className="small"
  data-testid="CourseCardDetails"
>
  provider-name
   • 
  test-course-number
   • access-message
   • 
  <Button
    className="m-0 p-0"
    onClick={[MockFunction useSelectSession.openSessionModal]}
    size="inline"
    variant="link"
  >
    change-or-leave-session-message
  </Button>
</span>
`;

exports[`CourseCard Details component has change session button on entitlement course but no access message 1`] = `
<span
  className="small"
  data-testid="CourseCardDetails"
>
  provider-name
   • 
  test-course-number
   • 
  <Button
    className="m-0 p-0"
    onClick={[MockFunction useSelectSession.openSessionModal]}
    size="inline"
    variant="link"
  >
    change-or-leave-session-message
  </Button>
</span>
`;
